File: Android/scenes/ktv_singbattle/src/main/java/io/agora/scene/ktv/singbattle/live/listener/SongActionListenerImpl.java
package io.agora.scene.ktv.singbattle.live.listener;

import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.LifecycleOwner;

import java.util.ArrayList;
import java.util.List;

import io.agora.scene.base.utils.LiveDataUtils;
import io.agora.scene.ktv.singbattle.live.RoomLivingViewModel;
import io.agora.scene.ktv.singbattle.service.RoomSelSongModel;
import io.agora.scene.ktv.singbattle.widget.song.OnSongActionListener;
import io.agora.scene.ktv.singbattle.widget.song.SongDialog;
import io.agora.scene.ktv.singbattle.widget.song.SongItem;

public class SongActionListenerImpl implements OnSongActionListener {
    private final LifecycleOwner mLifecycleOwner;
    private final RoomLivingViewModel mViewModel;
    private final boolean isChorus;

    public SongActionListenerImpl(
            LifecycleOwner activity,
            RoomLivingViewModel viewModel,
            boolean isChorus) {
        mLifecycleOwner = activity;
        mViewModel = viewModel;
        this.isChorus = isChorus;
    }

    @Override
    public void onChooseSongRefreshing(@NonNull SongDialog dialog) {
        // Choose song - list refresh
        LiveDataUtils.observerThenRemove(mLifecycleOwner, mViewModel.getSongList(), list -> {
            if (dialog.isVisible()) {
                dialog.setChooseRefreshingResult(transSongModel(list));
            }
        });
    }

    @Override
    public void onChooseSongChosen(@NonNull SongDialog dialog, @NonNull SongItem songItem) {
        // Choose song
        RoomSelSongModel songModel = songItem.getTag(RoomSelSongModel.class);
        LiveDataUtils.observerThenRemove(mLifecycleOwner, mViewModel.chooseSong(songModel), success -> {
            if (success && dialog.isVisible()) {
                dialog.setChooseSongItemStatus(songItem, true);
            } else if (!success) {
                songItem.loading = false;
                dialog.setChooseSongItemStatus(songItem, false);
            }
        });
    }

    @Override
    public void onChosenSongDeleteClicked(@NonNull SongDialog dialog, @NonNull SongItem song) {
        // Delete song
        RoomSelSongModel songModel = song.getTag(RoomSelSongModel.class);
        mViewModel.deleteSong(songModel);
    }

    @Override
    public void onChosenSongTopClicked(@NonNull SongDialog dialog, @NonNull SongItem song) {
        // Top song
        RoomSelSongModel songModel = song.getTag(RoomSelSongModel.class);
        mViewModel.topUpSong(songModel);
    }

    @Override
    public void onStartSingBattleGame(@NonNull SongDialog dialog) {
        mViewModel.startSingBattleGame();
    }

    public static List<SongItem> transSongModel(@Nullable List<RoomSelSongModel> data) {
        ArrayList<SongItem> list = new ArrayList<>();
        if (data != null) {
            for (RoomSelSongModel song : data) {
                SongItem item = new SongItem(
                        song.getSongNo(),
                        song.getSongName(),
                        song.getImageUrl(),
                        song.getSinger(),
                        song.getName(),
                        !TextUtils.isEmpty(song.getName()),
                        song.getUserNo()
                );
                item.setTag(song);
                list.add(item);
            }
        }
        return list;
    }
}

File: Android/scenes/cantata/src/main/java/io/agora/scene/cantata/live/listener/SongActionListenerImpl.kt
package io.agora.scene.cantata.live.listener

import android.text.TextUtils
import androidx.lifecycle.LifecycleOwner
import io.agora.scene.base.utils.LiveDataUtils
import io.agora.scene.cantata.live.RoomLivingViewModel
import io.agora.scene.cantata.service.RoomSelSongModel
import io.agora.scene.cantata.widget.song.OnSongActionListener
import io.agora.scene.cantata.widget.song.SongDialog
import io.agora.scene.cantata.widget.song.SongItem

/**
 * Song Selection Panel Listener
 */
class SongActionListenerImpl constructor(
    private val mLifecycleOwner: LifecycleOwner,
    private val mViewModel: RoomLivingViewModel,
    private val isChorus: Boolean
) : OnSongActionListener {
    override fun onChooseSongRefreshing(dialog: SongDialog) {
        // Song selection - list refresh
        LiveDataUtils.observerThenRemove(
            mLifecycleOwner,
            mViewModel.getSongList()
        ) { list: List<RoomSelSongModel>? ->
            if (dialog.isVisible) {
                dialog.setChooseRefreshingResult(transSongModel(list))
            }
        }
    }

    override fun onChooseSongChosen(dialog: SongDialog, songItem: SongItem) {
        // Select song
        val songModel = songItem.getTag(RoomSelSongModel::class.java) ?: return
        LiveDataUtils.observerThenRemove(
            mLifecycleOwner,
            mViewModel.chooseSong(songModel, isChorus)
        ) { success: Boolean ->
            if (success && dialog.isVisible) {
                dialog.setChooseSongItemStatus(songItem, true)
            } else if (!success) { // order failed
                songItem.loading = false
                dialog.setChooseSongItemStatus(songItem, false)
            }
        }
    }

    override fun onChosenSongDeleteClicked(dialog: SongDialog, song: SongItem) {
        // Delete song
        val songModel = song.getTag(RoomSelSongModel::class.java)?:return
        mViewModel.deleteSong(songModel)
    }

    override fun onChosenSongTopClicked(dialog: SongDialog, song: SongItem) {
        // Move to top
        val songModel = song.getTag(RoomSelSongModel::class.java)?:return
        mViewModel.topUpSong(songModel)
    }

    companion object {
        fun transSongModel(data: List<RoomSelSongModel>?): List<SongItem> {
            val list = ArrayList<SongItem>()
            if (data != null) {
                for (song in data) {
                    val item = SongItem(
                        song.songNo,
                        song.songName,
                        song.imageUrl,
                        song.singer,
                        song.name!!,
                        !TextUtils.isEmpty(song.name),
                        song.userNo
                    )
                    item.setTag(song)
                    list.add(item)
                }
            }
            return list
        }
    }
}